### DISCUSSION TITLE
uc.Remote for use in Docker container

### DISCUSSION THREAD

**Discussion by ajtadeo** [ultrafunkamsterdam/undetected-chromedriver#1600] [Ideas]
I'm trying to use Undetected Chrome driver within a Docker container, but there seems to be no Remote module to do so. This is the code I'm trying to use, as seen in this tutorial (https://www.browserstack.com/guide/run-selenium-tests-in-docker).

```python
options = uc.ChromeOptions()
options.add_argument('--headless=new')
options.add_argument("--disable-gpu")
options.add_argument('--blink-settings=imagesEnabled=false')

driver = uc.Remote(
    command_executor=f'http://{os.environ.get("SELENIUM_SERVICE_NAME")}:4444/wd/hub', 
    options=options, 
)
```

**Comment by matez0** [comment]
Did you managed to pass the cloudflare on https://nowsecure.nl/ as in the README example? - Me not.

But first, you have to build a patched remote selenium driver:
```dockerfile
# Dockerfile
FROM selenoid/chrome:117.0
USER root
RUN apt-get update || true
RUN apt-get install -y --no-install-recommends pip
RUN pip install undetected-chromedriver
COPY patch_driver.py .
RUN python3 patch_driver.py
```
and
```python
# patch_driver.py
from undetected_chromedriver import Patcher

Patcher(executable_path='/usr/bin/chromedriver').patch() or quit(1)
```
Then you can make a workaround:
```python
import undetected_chromedriver as uc


class RemoteService:
    SELENOID_HOST = 'your-selenoid-host'
    service_url = f'http://{SELENOID_HOST}:4444/wd/hub'  # If you use the driver container directly, remove `/wd/hub`.
    path = '/usr/bin/true'  # some existing fake path

    def start(self):
        pass

    def stop(self):
        pass


class UdChrome(uc.Chrome):
    def __init__(self):
        options = uc.ChromeOptions()
        # Set your options here...
        options._session = self
        
        super(uc.Chrome, self).__init__(options=options, service=RemoteService(), keep_alive=True)
        
        # Copy here what else you want from the uc.Chrome constructor...
        self._delay = 3
        self.options = options


# Test whether it passes the cloudflare check:
driver = UdChrome()
driver.get('https://nowsecure.nl')
import time; time.sleep(5)
driver.save_screenshot('nowsecure.png')
```
> Note: the code of undetected_chromedriver is a bit chaotic.

**Reply by ajtadeo** [reply]
thanks for the reply! 😄 i will try this with my app and see if it works. the code i gave in my initial post was not able to pass [https://nowsecure.nl/](https://nowsecure.nl/) because i didn't know how to implement the remote service so UC was failing. hopefully this can be a feature added in the future so developers don't have to do a workaround.

**Reply by ajtadeo** [reply]
@matez0 for your `path_driver.py`, you set the path to the chromedriver for patcher. does the patch override the source of the chromedriver to not use this path?

**Reply by matez0** [reply]
The path in the `patch_driver.py` is the default path of the binary executable in the selenoid/chrome image.

The `patch` will replace some byte patterns in the original executable file `/usr/bin/chromedriver` for anti bot detection.

The `/usr/bin/chromedriver` is a server application which is listening on the configured port (4444) for selenium commands as you can see in the [entrypoint.sh](https://github.com/aerokube/images/blob/master/static/chrome/entrypoint.sh). The `uc.Chrome` based class will communicate with the patched server application via the endpoint configured in `RemoteService`.

The docker file given by me would be more proper, if it would switch back the user with `USER selenium` as you can see in [the base docker file](https://github.com/aerokube/images/blob/5a2dd3430a60ceb7931f2a785ba2a52112b8b8fa/static/chrome/Dockerfile#L25C1-L25C14), but it is suitable for testing now.

**Reply by rchvalbo** [reply]
@ajtadeo  any luck with this solution? Was hoping to see if you had any luck before I took a stab at it. If you haven't tried yet I'll probably try this out soon.

**Reply by matez0** [reply]
There is an issue with inheriting from `uc.Chrome`. It is better to inherit from normal selenium `Chrome` and picking / calling some methods from `uc.Chrome` class.

What I found useful so far is only the patching. One can make a multi-stage build for the patching as well patching the binary in a `python:alpine` image and copying back the patched file into the `selenoid/chrome` image to avoid running the `apt` and installing extra packages that are not used runtime.

**Reply by matez0** [reply]
What I would like to know that could someone reproduce to pass the cloudflare on https://nowsecure.nl/ as in the README example?

**Reply by hnnoye04** [reply]
@matez0 https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/1577 prob try with chrome 116?

**Reply by matez0** [reply]
Using this docker file:
```Dockerfile
# Dockerfile
ARG CHROME_VERSION

FROM selenoid/chrome:${CHROME_VERSION:-latest} as selenoid_chrome


FROM python:3.10-alpine as patcher

RUN pip install undetected-chromedriver

COPY --from=selenoid_chrome /usr/bin/chromedriver /usr/bin/chromedriver

COPY patch_driver.py .

RUN python3 patch_driver.py


FROM selenoid_chrome

USER root

COPY --from=patcher /usr/bin/chromedriver /usr/bin/chromedriver

ENV DRIVER_ARGS="--headless=new --user-agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36' --verbose"

RUN AAA=$(echo '#!/bin/bash\necho $DRIVER_ARGS' ; cat /entrypoint.sh ; ) && echo "$AAA" > /entrypoint.sh

USER selenium
```
where the user agent option from #1577 and the headless option are explicitly added to the driver arguments

and with the patch driver script from [this comment](https://github.com/ultrafunkamsterdam/undetected-chromedriver/discussions/1600#discussioncomment-7253842)

I built a chrome driver 116 image:
```bash
docker build -t ud-chrome-116 --build-arg=CHROME_VERSION="116.0" .
```
Whether the `DRIVER_ARGS` is passed to the driver can also be checked:
```bash
echo "tail /entrypoint.sh" | docker run --rm -i -p 127.0.0.1:4444:4444 --entrypoint=/bin/bash ud-chrome-116
```

The following test script extends the undetectable chrome driver to be able to use a remote chrome driver:
```python
# test_uc.py
import undetected_chromedriver as uc


class RemoteService:
    SELENOID_HOST = '127.0.0.1'
    service_url = f'http://{SELENOID_HOST}:4444'
    path = '/usr/bin/true'

    def start(self):
        pass

    def stop(self):
        pass


class Uchrome(uc.Chrome):
    def __init__(self, options):
        options._session = self
        super(uc.Chrome, self).__init__(options=options, service=RemoteService(), keep_alive=True)
        self._delay = 3
        self.options = options


options = uc.ChromeOptions()
options.headless=True

ua = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36'
options.add_argument(f'--user-agent={ua}')
options.add_argument("--headless=new")

driver = Uchrome(options)
driver._configure_headless()
driver.get('https://nowsecure.nl')
import time; time.sleep(5)
driver.save_screenshot('nowsecure.png')
```
Instead of calling the original constructor, I try to set everything to configure headless and also set the user agent option from #1577.

Starting the container:
```bash
docker run --rm -p 127.0.0.1:4444:4444 ud-chrome-116
```
And running the test script, the captured screenshot shows that the cloudflare is not passed.

@hnnoye04, you can execute these steps and see what happens.

**Comment by hnnoye04** [comment]
@matez0 can your one pass iphey/pixelscan? If not maybe it no using the ucdriver

**Reply by matez0** [reply]
With these additional options to get bigger screenshot:
```python
options.add_argument("--window-size=1024,1620")
options.add_argument("--start-maximized")
```

my result for [that site](https://iphey.com/):

![uc-screen](https://github.com/ultrafunkamsterdam/undetected-chromedriver/assets/29123073/156009b5-e0f1-4f01-830d-dc5f47816fd1)

You can see that the site detected chrome 117 because it was given in the user agent option, but I used chrome driver 116.
If I changed the chrome version in the user agent option, then the site display the changed chrome version.

**Comment by hnnoye04** [comment]
<img width="913" alt="image" src="https://github.com/ultrafunkamsterdam/undetected-chromedriver/assets/79848180/0417096e-1e17-4fe0-bee3-6f8322e8bab7">

**Comment by matez0** [comment]
I could not reproduce so far what undetected chromedriver states in the readme.

@hnnoye04 Could you provide a docker file to reproduce the stated feature for passing cloudflare on https://nowsecure.nl?

**Comment by hnnoye04** [comment]


**Reply by matez0** [reply]
Could you provide a docker file with which this script gives the expected result?

**Reply by hnnoye04** [reply]
This should work fine with your docker file I test it

**Reply by matez0** [reply]
Yes. It results the passed screenshot now. :)

If the selenium has this feature now, do we still need the undetected chrome driver at all?

**Reply by hnnoye04** [reply]
Without undetected driver you will get block by most bot detection website with this driver you can hide your fingerprint and auto frame work

**Reply by ajtadeo** [reply]
@matez0 do you mind posting the code snippet/instructions for how you implemented the solution? glad to see a workaround was found!

**Reply by hnnoye04** [reply]
@ajtadeo build the docker then take a look at seleniumbase patch don’t want to get by cf soon

**Reply by matez0** [reply]
@ajtadeo, @hnnoye04
The workaround is not only relying on the undetected-chromedriver, but uses some code from https://github.com/seleniumbase/SeleniumBase/commit/be477faa56c583c62defe957e8181ff1c43d8307. The SeleniumBase has MIT license and it would be good to know how much one can reuse from their solution which here is not so much.

The `open` or `get` method of the `undetected_chromedriver.Chrome` class could be implemented like `seleniumbase.core.browser_launcher.uc_open` function.

I experimented with the driver options suggested by @hnnoye04 and only these options were really necessary to pass the cloudflare:
```python
options.add_argument('--disable-blink-features=AutomationControlled')
with open('ext.crx', 'wb') as f:
    f.write(requests.get('https://nopecha.com/f/ext.crx').content)
options.add_extension('ext.crx')
```
these options were **not** necessary:
```python
options.set_capability("selenoid:options", {"enableVNC": True})
options.add_argument("--start-maximized")
options.add_argument('--disable-features=TranslateUI')
options.add_argument('--disable-translate')
options.add_argument('--lang=en-US')
options.add_argument('--disable-web-security')
options.add_argument("--no-sandbox");
options.add_argument("--disable-popup-blocking")
options.add_argument("--disable-notifications")
options.add_argument('--disable-dev-shm-usage')
options.add_experimental_option("excludeSwitches", ['enable-automation'])
options.add_experimental_option('useAutomationExtension', False)
ua = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36'
options.add_argument(f'--user-agent={ua}')
options.add_argument("--headless=new")
```
I used the chrome driver version 116 and even removed the setting of `DRIVER_ARGS` from the docker file mentioned [here](https://github.com/ultrafunkamsterdam/undetected-chromedriver/discussions/1600#discussioncomment-7864016).

We do not need the `undetected_chromedriver.Chrome` class neither.
Only the patched selenoid image from [here](https://github.com/ultrafunkamsterdam/undetected-chromedriver/discussions/1600#discussioncomment-7864016).

> **Note:** To pass the cloudflare, the [nopecha](https://nopecha.com) chrome extension was essential. Without this extension and the special window opening based on the SeleniumBase, using the patched chrome driver and the `undetected_chromedriver.Chrome` class cannot pass the cloudflare as stated in the README.

Finally, the reduced script:
```python
import time

import requests
from selenium.webdriver.chrome.remote_connection import ChromeRemoteConnection
from selenium.webdriver.chromium.options import ChromiumOptions
from selenium.webdriver.remote.webdriver import WebDriver

SELENOID_HOST = 'localhost'
executor = ChromeRemoteConnection(remote_server_addr=f'http://{SELENOID_HOST}:4444')

options = ChromiumOptions()

options.add_argument('--disable-blink-features=AutomationControlled')

with open('ext.crx', 'wb') as ext_file:
    ext_file.write(requests.get('https://nopecha.com/f/ext.crx').content)
options.add_extension('ext.crx')

driver = WebDriver(command_executor=executor, options=options)

url = 'https://nowsecure.nl'
driver.execute_script(f'window.open("{url}","_blank");')
time.sleep(3)
driver.execute_script("window.close();")
driver.switch_to.window(driver.window_handles[-1])
time.sleep(8)
driver.save_screenshot('nowsecure.png')
```
Of course, you can use `Uchrome` class from [here](https://github.com/ultrafunkamsterdam/undetected-chromedriver/discussions/1600#discussioncomment-7864016) or similar instead of `WebDriver`, if you would like to use something from the `undetected_chromedriver.Chrome` class as well.

**Reply by hnnoye04** [reply]
@ajtadeo  if you wish to use the extension follow this guide https://developers.nopecha.com/guides/extension/
and test on  any on these https://nopecha.com/demo

**Comment by hnnoye04** [comment]
@matez0  Just like thissss i just copy the code from https://github.com/seleniumbase/SeleniumBase/commit/be477faa56c583c62defe957e8181ff1c43d8307 xddd

**Comment by Tongcheng** [comment]
Doesn't work for me when using the above window open/close tactic with nopecha, does it still work for you?

**Reply by hnnoye04** [reply]
wdym any code?

**Reply by Tongcheng** [reply]
Thanks for the fast reply @hnnoye04 ! I am not using a selenoid docker, but using a selenium docker, then manually patched to remove the "fingerprint" (cdc). 

With residential proxy, this doesn't seems to pass the hardware fingerprint of pixelscan.net (), also for nowsecure.nl, it also doesn't pass.

Pixelscan 1:
![fingerprint1](https://github.com/ultrafunkamsterdam/undetected-chromedriver/assets/6565047/b5454756-6010-42fe-9048-808aef2543b9)

Pixelscan 2:
![fingerprint2](https://github.com/ultrafunkamsterdam/undetected-chromedriver/assets/6565047/0ab75a2f-cdfa-47a2-b8ff-09173ad04efb)

Nowsecure.nl: 
![nowsecure](https://github.com/ultrafunkamsterdam/undetected-chromedriver/assets/6565047/65838465-070c-4e98-a7d8-63b33e4753ac)


My code:
`
class RemoteChromeDriver:
    def __init__(self, user_profile=None, proxy_str=None, proxy_auth=None, extension_path=None):
        self.chrome_options = Options()
        self.desired_capabilities = DesiredCapabilities.CHROME
        self.chrome_options.add_argument('--remote-debugging-pipe')
        self.chrome_options.add_argument("--start-maximized")
        self.chrome_options.add_argument("--no-sandbox")
        self.chrome_options.add_argument("--disable-dev-shm-usage")
        self.chrome_options.add_argument('--disable-blink-features=AutomationControlled')
        self.chrome_options.add_argument("--disable-browser-side-navigation")
        # More stealth - against CloudFlare 
        self.chrome_options.add_argument("start-maximized")
        self.chrome_options.add_experimental_option("excludeSwitches", ["enable-automation"])
        self.chrome_options.add_experimental_option('useAutomationExtension', False)
        <... some other logic... >
        # Try NoPecha: https://github.com/ultrafunkamsterdam/undetected-chromedriver/discussions/1600
        with open('ext.crx', 'wb') as ext_file:
            ext_file.write(requests.get('https://nopecha.com/f/ext.crx').content)
        self.chrome_options.add_extension('ext.crx')
        
    def __enter__(self):
        grant_ip_to_hub_security()
        self.driver = webdriver.Remote(
            command_executor=f'{hub_url}/wd/hub',
            desired_capabilities=self.desired_capabilities,
            options=self.chrome_options
        )
        return self.driver
`

Running it according to above example:
`import time

url = 'https://nowsecure.nl'
driver.execute_script(f'window.open("{url}","_blank");')
time.sleep(3)
driver.execute_script("window.close();")
driver.switch_to.window(driver.window_handles[-1])
time.sleep(8)
driver.save_screenshot('nowsecure.png')`

Any thoughts what might be causing the cloudflare fail to pass here?

**Reply by hnnoye04** [reply]
try use nodriver or seleniumbase directly this repo no update again more.This version of ud driver detect already

**Reply by hnnoye04** [reply]
```
from seleniumbase import SB

def open_the_turnstile_page(sb):
    sb.driver.uc_open_with_reconnect(
        "https://nowsecure.nl/", reconnect_time=2.5,
    )

def click_turnstile_and_verify(sb):
    sb.driver.uc_switch_to_frame("iframe")
    sb.driver.uc_click("span.mark")
    sb.assert_element("img#captcha-success", timeout=3.33)

with SB(uc=True, test=True) as sb:
    open_the_turnstile_page(sb)
    try:
        click_turnstile_and_verify(sb)
    except Exception:
        open_the_turnstile_page(sb)
        click_turnstile_and_verify(sb)
    sb.set_messenger_theme(location="top_left")
    sb.post_message("SeleniumBase wasn't detected", duration=3)
```
cause uc_open_with_reconnect work with seleniumbase driver

**Reply by Tongcheng** [reply]
Appreciate the help and directions @hnnoye04 ! Let me try both!

**Reply by big-camel** [reply]
```py
driver.execute_cdp_cmd("Page.addScriptToEvaluateOnNewDocument", {
    "source": """
      var newProto = navigator.__proto__;
      delete newProto.webdriver;
      newProto.permissions={query:function(parameters){ return parameters.name === 'notifications' ? Promise.resolve({ state: Notification.permission }) : originalQuery(parameters) }};
      newProto.language='en';newProto.languages=['zh-CN', 'zh', 'en'];
      navigator.__proto__ = newProto;
    """
  })
```
Some applications may use JavaScript to detect the webdriver property

**Reply by Tongcheng** [reply]
Thank you @big-camel ! Will try it out!
